### ISSUE TITLE
Bots sometimes don't start

### ISSUE THREAD

**Issue by certicky** [certicky/sscait-tournamentmanager#2]
As we've noticed during 2021/22 round-robin phase, some bots sometimes don't start on the client machine.
This happens quite rarely and it doesn't seem to affect specific bots, but it's a problem anyway, because it affects the results.

Potential solution:
* The Java Client could actually make sure that its bot has indeed been started when the `Client_RunProxyScript` function was called.
* The Client should wait some time and periodically check if the bot's Windows process is indeed running. The `WindowsCommandTools.GetRunningProcesses()` could probably be used for this.
* Note: There already is something like this on line 83 of `/client/ClientCommands.java`, but it doesn't seem to work well enough.

https://github.com/certicky/sscait-tournamentmanager/blob/ad74cfba360446635e055c41e7717eb77635cb3b/client/ClientCommands.java#L83

**Comment by layland34** [comment]
Shouldn't line 86 be < instead of >= ?
That's if i understand the comment above correctly.

**Comment by certicky** [comment]
> Shouldn't line 86 be < instead of >= ? That's if i understand the comment above correctly.

I don't think so. The variable name "nowRunning" is a bit stupid, but the logic of that function is this:
- At the beginning, store the number of currently running processes in "nowRunning".
- And then use the while loop to keep trying to run the bot until there are more running processes than there were in the beginning.

In other words, "while no. of running processes isn't biger than _nowRunning_, try running the bot".

That variable should be renamed to "runningAtTheBeginning" or something like that to make it less confusing.

**Comment by layland34** [comment]
Ok, that's what i thought. But then it needs to be reversed.
Unless there are actually less processes when the bot is running.

**Comment by chriscoxe** [comment]
As discussed on Discord, I got all the proxy bots that are enabled on SSCAIT running in my environment, and a handful of DLL bots, but wasn't able to reproduce the problem. I've changed a couple of things to work more like https://github.com/davechurchill/StarcraftAITournamentManager. I don't know whether my changes will help.

I got rid of the `while` loop at:
```
		Vector<Integer> nowRunning = WindowsCommandTools.GetRunningProcesses(); 
		int tries = 0;
		while (tries < 5 && nowRunning.size() >= WindowsCommandTools.GetRunningProcesses().size() ) {
```
because if the total number of processes has increased but `tries` is still 0, it won't run `bwapi-data/AI/run_proxy.bat` at all (not even once) and because https://github.com/davechurchill/StarcraftAITournamentManager only runs `bwapi-data/AI/run_proxy.bat` once (it's possible that if unrelated processes were created/terminated during this while loop, multiple instances of the bot could have been started concurrently and might interfere with each other). I'm guessing the original rationale might have been if there are bots that self-extract themselves when they are run, e.g. Java may unzip DLL file dependencies from JAR files that contain the dependencies. All enabled SSCAIT proxy bots and CherryPiSSCAIT2017 still seem to work for me after my changes.

https://github.com/davechurchill/StarcraftAITournamentManager uses the `/T` flag when killing Starcraft, which causes it to also kill any child processes that were started by it. Presumably it was added for a reason, and I doubt it could hurt to use it.

As requested, I've pushed the new JAR files to GitHub (I'm using `jdk-8u231-windows-i586.exe`). I'm a bot author, so for credibility, or if you're using an earlier version of Java on SSCAIT, you might want to overwrite my JAR files.

**Comment by chriscoxe** [comment]
The issue was/is happening on SSCAIT for all bots, including DLL bots. Looking at BWAPI's source code, this can happen if a proxy bot from an old game is still running for some reason when the game starts. In this scenario, the proxy bot connects to the game, and the DLL bot is never loaded. The proxy bot might not do anything (i.e. not issue any commands), making it a no-start game, if its race is not what it expects, or e.g. if it crashes on frame 0 because it can't find its config files. So I think what was/is happening is that bot instance(s) from old games are causing the no-start issue. If this theory is correct, I expect there are replays where a bot played like a different bot! Theoretically, it could also corrupt the bots read/write files if it uses similar filenames.

If the no-start issue still happens after my changes are used, the first thing to investigate would be to see whether any proxy bots from old games still running, e.g. extra java processes (i.e. in addition to the sscait-tournamentmanager's single client java process) or bot EXE processes. For the currently enabled SSCAIT bots, they would be at least, in alphabetical order: adias.exe, Bereaver.exe, Boris.exe, CactusClient.exe, CUNYAIBot.exe, java.exe (more than 1 if a DLL bot should currently be running, or more than 2 if a JAVA_MIRROR or JAVA_JNI bot should currently be running), Monster.exe, StarterBot.exe, strategy.exe (it's spawned by adias), styx_z.exe, tsc-bwai.exe, UAlbertaBot.exe, ZergHell.exe.

**Comment by certicky** [comment]
OK, great. I just deployed the version from this commit: https://github.com/certicky/sscait-tournamentmanager/commit/2fe47ca763969876642d6177999bc7ef89a9cad5 

So let's wait some time, play some games, and then we'll check if there were still some no-start games.

**Comment by chriscoxe** [comment]
Ok. If you haven't already, please restart the client VMs, in case bot(s) from old games are still running.
